File: bot/strategies/homepage.js
import fetch from 'node-fetch';
import * as cheerio from 'cheerio';

const UAs = [
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
];

function ua() { return UAs[Math.floor(Math.random() * UAs.length)]; }

const DEAL_SELECTORS = [
  '[class*="announcement"]', '[class*="announce"]', '[class*="banner"]',
  '[class*="promo"]', '[class*="offer"]', '[class*="deal"]',
  '[class*="discount"]', '[class*="coupon"]', '[class*="sale"]',
  '[class*="hero"]', '[class*="cta"]', '[class*="special"]',
  '[id*="announcement"]', '[id*="promo"]', '[id*="offer"]',
  '[aria-label*="announcement"]', '[aria-label*="promo"]',
];

const DEAL_WORDS = [
  'free', 'save', 'off', '% off', 'discount', 'coupon', 'promo',
  'deal', 'offer', 'sale', 'limited time', 'special offer',
  'get started', 'sign up', 'try free', 'start free',
  'refer', 'referral', 'invite', 'bonus', 'credit',
];

export async function scanHomepage(brand) {
  const { brandName, website } = brand;
  const url = website.replace(/\/$/, '');
  const results = [];

  try {
    const res = await fetch(url, {
      headers: { 'User-Agent': ua() },
      timeout: 10000,
      redirect: 'follow',
    });
    if (!res.ok) return results;

    const html = await res.text();
    const $ = cheerio.load(html);
    const text = $('body').text().toLowerCase();

    // Check for deal announcement elements
    const dealElements = [];
    for (const sel of DEAL_SELECTORS) {
      $(sel).each((_, el) => {
        const t = $(el).text().trim();
        if (t.length > 5 && t.length < 200) dealElements.push(t);
      });
    }

    // Check for common deal patterns in page text
    const pctMatch = text.match(/(\d+%)\s*off/i);
    const dollarMatch = text.match(/\$(\d+)\s*off/i);
    const freeMatch = text.match(/free\s+(trial|shipping|delivery)/i);

    let confidence = 0;
    let discount = '';
    let title = '';

    if (dealElements.length > 0) {
      confidence = 60 + Math.min(dealElements.length * 5, 30);
      title = dealElements[0].slice(0, 200);
      if (pctMatch) discount = pctMatch[0];
      else if (dollarMatch) discount = dollarMatch[0];
      else if (freeMatch) discount = freeMatch[0];
      else discount = 'Special offer';
    } else if (pctMatch || dollarMatch || freeMatch) {
      confidence = 40;
      discount = pctMatch?.[0] || dollarMatch?.[0] || freeMatch?.[0] || '';
      title = `${brandName} — ${discount}`;
    }

    if (confidence > 0) {
      results.push({
        sourceUrl: url,
        sourcePage: 'homepage',
        sourceReliability: 'Official Site',
        confidence,
        title: title.slice(0, 200),
        description: $('meta[name="description"]').attr('content')?.slice(0, 500) || '',
        discount: discount || 'Check homepage for offers',
        codes: [],
      });
    }

    // Also check /blog for deal-related posts
    try {
      const blogRes = await fetch(`${url}/blog`, {
        headers: { 'User-Agent': ua() },
        timeout: 8000,
        redirect: 'follow',
      });
      if (blogRes.ok) {
        const blogHtml = await blogRes.text();
        const blog$ = cheerio.load(blogHtml);
        const blogText = blog$.text().toLowerCase();
        if (DEAL_WORDS.some(w => blogText.includes(w))) {
          results.push({
            sourceUrl: `${url}/blog`,
            sourcePage: 'blog',
            sourceReliability: 'Official Site',
            confidence: 50,
            title: `${brandName} Blog — possible deals`,
            description: 'Blog contains deal-related keywords',
            discount: 'Check blog for offers',
            codes: [],
          });
        }
      }
    } catch {}

  } catch {}

  return results;
}

File: bot/app/adapters/homepage.py
import re
from bs4 import BeautifulSoup

UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) Chrome/120.0.0.0'
SELECTORS = ['[class*="announcement"]','[class*="promo"]','[class*="offer"]','[class*="deal"]','[class*="discount"]','[class*="coupon"]','[class*="sale"]']
DEAL_WORDS = ['free','save','off','% off','discount','coupon','promo','deal','offer','sale','limited time','special offer']

class HomepageAdapter:
    name = "homepage"

    async def discover(self, brand: dict, client):
        url = brand['website'].rstrip('/')
        bn = brand['brandName']
        results = []
        try:
            r = await client.get(url, timeout=4.0)
            if r.status_code != 200: return results
            soup = BeautifulSoup(r.text, 'lxml')
            text = soup.get_text().lower()
            deal_els = []
            for sel in SELECTORS:
                for el in soup.select(sel):
                    t = el.get_text(strip=True)
                    if 5 < len(t) < 200: deal_els.append(t)
            pct = re.search(r'(\d+%)\s*off', text)
            dollar = re.search(r'\$(\d+)\s*off', text)
            free = re.search(r'free\s+(trial|shipping|delivery)', text)
            conf = 0
            discount = ''
            title = ''
            if deal_els:
                conf = 60 + min(len(deal_els) * 5, 30)
                title = deal_els[0][:200]
                discount = (pct or dollar or free).group(0) if (pct or dollar or free) else 'Special offer'
            elif pct or dollar or free:
                conf = 40
                discount = (pct or dollar or free).group(0)
                title = f"{bn} — {discount}"
            if conf > 0:
                meta = soup.find('meta', attrs={'name': 'description'})
                desc = meta.get('content', '')[:500] if meta else ''
                results.append({"sourceUrl": url, "sourcePage": "homepage", "confidence": conf,
                                "title": title[:200], "description": desc, "discount": discount or 'Check homepage',
                                "codes": [], "blocked": False})
            try:
                blog_r = await client.get(f"{url}/blog", timeout=3.0)
                if blog_r.status_code == 200:
                    blog_soup = BeautifulSoup(blog_r.text, 'lxml')
                    blog_text = blog_soup.get_text().lower()
                    if any(w in blog_text for w in DEAL_WORDS):
                        results.append({"sourceUrl": f"{url}/blog", "sourcePage": "blog", "confidence": 50,
                                        "title": f"{bn} Blog — possible deals", "description": "",
                                        "discount": "Check blog for offers", "codes": [], "blocked": False})
            except: pass
        except: pass
        return results
